### PR TITLE
[FullscreenReminder] Add "Show full screen reminder to press Esc on exit" option in settings page

### PR DESCRIPTION
**Pull request by jagadeshjai**
- Add an option to Show/Hide `fullscreen reminder to press Esc on exit` message when going fullscreen on videos.
- Add feature flag for the same which is disabled by default.

Note: This PR does not intend to hide the `To exit fullscreen, press F11` bubble from this preference.

Resolves https://github.com/brave/brave-browser/issues/38928

Outcome:
image

https://github.com/brave/brave-core/assets/45998266/0fad3302-4f1b-43e8-af84-cca38404fe49

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used GitHub [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
See https://github.com/brave/brave-browser/issues/38928

### CODE REVIEW

**Comment by bsclifton** [comment]
@jagadeshjai I think you forgot a commit or forgot to add files to the commit- I'm only seeing the feature being added, not the setting or the handling of the feature

**Comment by jagadeshjai** [comment]
> @jagadeshjai I think you forgot a commit or forgot to add files to the commit- I'm only seeing the feature being added, not the setting (ex: in brave://settings/system) or the handling of the feature

@bsclifton Its actually in the draft, yet to do that part.

**Comment by jagadeshjai** [comment]
@bsclifton  This change is ready for review now, PTAL.

**Comment by bsclifton** [comment]
Pulled down the code and tried it - works great! 😄 (just had to add one commit- compile had failed for me)
I'll help code review it and also find the other folks who can help review

Huge thanks for grabbing this. Stay tuned please

File: patches/chrome-browser-ui-views-frame-browser_view.cc.patch [hunk 1]
```diff
@@ -0,0 +1,15 @@
+diff --git a/chrome/browser/ui/views/frame/browser_view.cc b/chrome/browser/ui/views/frame/browser_view.cc
```
Comment (emerick) [inline patches/chrome-browser-ui-views-frame-browser_view.cc.patch:1]: I think this patch was added by mistake?

File: app/brave_settings_strings.grdp [hunk 2]
```diff
@@ -1473,6 +1473,9 @@
   <message name="IDS_SETTINGS_CLOSING_LAST_TAB_OPTION_LABEL" desc="The text for settings option">
     Close window when closing last tab
   </message>
+  <message name="IDS_SETTINGS_SHOW_FULLSCREEN_REMINDER_OPTION_LABEL" desc="The text for fullscreen reminder settings option">
+    Show full screen reminder to press Esc on exit
```
Comment (emerick) [inline app/brave_settings_strings.grdp:1477]: Hmm, when I enter fullscreen mode on Windows the message says "To exit full screen, press F11". It might be better to reword this as "Show exit instructions when entering full screen mode".

File: browser/brave_browser_features.h [hunk 3]
```diff
@@ -17,8 +17,9 @@ BASE_DECLARE_FEATURE(kBraveCleanupSessionCookiesOnSessionRestore);
 BASE_DECLARE_FEATURE(kBraveCopyCleanLinkByDefault);
 BASE_DECLARE_FEATURE(kBraveOverrideDownloadDangerLevel);
 BASE_DECLARE_FEATURE(kBraveWebViewRoundedCorners);
-
 BASE_DECLARE_FEATURE(kBraveDayZeroExperiment);
+BASE_DECLARE_FEATURE(kBraveShowFullscreenReminder);
```
Comment (emerick) [inline browser/brave_browser_features.h:21]: Maybe rename to `kBraveShowFullscreenExitInstructions` for clarity?

File: browser/brave_profile_prefs.cc [hunk 4]
```diff
@@ -461,6 +461,7 @@ void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry) {
                                 base::Value(true));
   registry->RegisterBooleanPref(kEnableWindowClosingConfirm, true);
   registry->RegisterBooleanPref(kEnableClosingLastTab, true);
+  registry->RegisterBooleanPref(kEnableShowFullscreenReminder, true);
```
Comment (emerick) [inline browser/brave_profile_prefs.cc:464]: Maybe rename to `kEnableShowFullscreenExitInstructions` for clarity?

File: browser/resources/settings/brave_overrides/system_page.ts [hunk 5]
```diff
@@ -105,6 +105,19 @@ RegisterPolymerTemplateModifications({
     )
     // </if>
 
+    if(loadTimeData.getBoolean("isShowFullscreenReminderEnabled")) {
```
Comment (emerick) [inline browser/resources/settings/brave_overrides/system_page.ts:108]: Nit: Space after the `if` keyword.

File: browser/ui/webui/settings/brave_settings_localized_strings_provider.cc [hunk 6]
```diff
@@ -400,6 +400,8 @@ void BraveAddCommonStrings(content::WebUIDataSource* html_source,
       {"braveClosingLastTab", IDS_SETTINGS_CLOSING_LAST_TAB_OPTION_LABEL},
       {"braveDisableClickableMuteIndicators",
        IDS_SETTINGS_DISABLE_CLICKABLE_MUTE_INDICATORS},
+      {"braveShowFullscreenReminder",
+       IDS_SETTINGS_SHOW_FULLSCREEN_REMINDER_OPTION_LABEL},
```
Comment (emerick) [inline browser/ui/webui/settings/brave_settings_localized_strings_provider.cc:404]: Maybe rename to `IDS_SETTINGS_SHOW_FULLSCREEN_EXIT_INSTRUCTIONS` for clarity?

File: chromium_src/chrome/browser/ui/views/frame/browser_view.cc [hunk 7]
```diff
@@ -58,3 +62,20 @@ void BrowserView::SetNativeWindowPropertyForWidget(views::Widget* widget) {
...
   widget->SetNativeWindowProperty(kBrowserViewKey, this);
 }
+
+void BrowserView::UpdateExclusiveAccessBubble(
+    const ExclusiveAccessBubbleParams& params,
+    ExclusiveAccessBubbleHideCallback first_hide_callback) {
+  // Show/Hide full screen reminder bubble based on our settings preference
+  // for tab-initiated ones.
+  if (base::FeatureList::IsEnabled(features::kBraveShowFullscreenReminder)) {
+    if (!GetProfile()->GetPrefs()->GetBoolean(kEnableShowFullscreenReminder) &&
```
Comment (emerick) [inline chromium_src/chrome/browser/ui/views/frame/browser_view.cc:72]: Could probably be condensed into a single if statement.

File: app/brave_settings_strings.grdp (hunk 2 above)
Comment (jagadeshjai) [reply app/brave_settings_strings.grdp:1477]: Hi @emerick 

We actually intend to hide only the `Show fullscreen reminder to press Esc to exit`, which pops up only when the video goes into fullscreen mode (basically tab-initiated fullscreen). Hence, we will show `To exit fullscreen, press F11` regardless of this preference.

FYR: https://github.com/brave/brave-browser/issues/38928#issuecomment-2166653019

I think renaming the PR title and description will avoid this confusion.

File: chromium_src/chrome/browser/ui/views/frame/browser_view.cc (hunk 7 above)
Comment (jagadeshjai) [reply chromium_src/chrome/browser/ui/views/frame/browser_view.cc]: Done.

File: browser/resources/settings/brave_overrides/system_page.ts (hunk 5 above)
Comment (jagadeshjai) [reply browser/resources/settings/brave_overrides/system_page.ts]: Done.

File: patches/chrome-browser-ui-views-frame-browser_view.cc.patch (hunk 1 above)
Comment (jagadeshjai) [reply patches/chrome-browser-ui-views-frame-browser_view.cc.patch]: Removed.

Comment (bsclifton) [reply patches/chrome-browser-ui-views-frame-browser_view.cc.patch]: OMG - sorry about this 😓

**Comment by bsclifton** [comment]
While it's a nice touch to have the feature on brave://flags, I think this is a pretty safe feature and we'd have to remove the flag eventually.

it's OK to have under brave://settings/system as-is. Can you please remove the entry on brave://flags? 😄

**Comment by jagadeshjai** [comment]
> While it's a nice touch to have the feature on brave://flags, I think this is a pretty safe feature and we'd have to remove the flag eventually.
> 
> it's OK to have under brave://settings/system as-is. Can you please remove the entry on brave://flags? 😄

Sure @bsclifton ! Will do it now.

File: app/brave_settings_strings.grdp (hunk 2 above)
Comment (emerick) [reply app/brave_settings_strings.grdp:1477]: I see, I didn't realize that was a separate mode.

File: browser/brave_profile_prefs.cc (hunk 4 above)
Comment (emerick) [reply browser/brave_profile_prefs.cc:464]: Actually, I don't think we need the word `Enable` in there at all. `Show` is the action (we're either going to show the instructions or not).

Comment (jagadeshjai) [reply browser/brave_profile_prefs.cc]: Done.

Renamed to `kShowFullscreenReminder`.


Do you want me to change to `kShowFullscreenExitInstruction`, If yes, I need to change the String in `.grd` as well, Kindly confirm.

Comment (emerick) [reply browser/brave_profile_prefs.cc]: I think `kShowFullscreenExitInstructions` is a little clearer, but it's up to you.

**Review by emerick — APPROVED** [review]

**Review by bsclifton — APPROVED** [review]
Really nice work on this one! Pulled it down, retested.
Code changes look great too 😄 

I rebased and added a test plan to the original issue. Next step is to let CI run - then we're good for merging!

Thanks again, @jagadeshjai! ❤️

**Comment by bsclifton** [comment]
Created https://github.com/brave/brave-core/pull/24236 to allow for a full run of CI - once that passes, we're good to merge!

**Comment by bsclifton** [comment]
CI looks great - merging 😄

**Comment by jagadeshjai** [comment]
Thanks @bsclifton  :heart_decoration: :smile:
